### PR TITLE
[+] Permissions - Add a permission mechanism to protect the data accordingly to the UI configuration

### PR DESCRIPTION
**Pull request by ArnaudValensi**
## Pull Request checklist:

- [ ] Write an explicit title for the Pull Request
- [ ] Write changes made in the CHANGELOG.md
- [ ] Create automatic tests
- [ ] No automatic tests failures
- [ ] Test manually the implemented changes
- [ ] Review my own code (indentation, syntax, style, simplicity, readability)
- [ ] Wonder if you can improve the existing code

### CODE REVIEW

**Comment by ArnaudValensi** [comment]
[(5) Automated Permissions - Rails liana](https://trello.com/c/AAtSaH4V/2568-5-automated-permissions-rails-liana)

File: app/services/forest_liana/permissions_checker.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+module ForestLiana
+  class PermissionsChecker
+    @@permissions = nil
+    @@last_retrieve = nil
+
+    def initialize(resource, permission_name)
+      @collection_name = ForestLiana.name_for(resource)
+      @permission_name = permission_name
+
+      @expiration_in_seconds = !ENV['FOREST_PERMISSIONS_EXPIRATION_IN_SECONDS'].nil? ?
```
Comment (arnaudbesnier) [inline app/services/forest_liana/permissions_checker.rb]:
```ruby
@expiration_in_seconds = ENV['FOREST_PERMISSIONS_EXPIRATION_IN_SECONDS'] || 3600
```

Comment (arnaudbesnier) [reply app/services/forest_liana/permissions_checker.rb]: This should be more a class variable than an instance variable by the way.

File: app/services/forest_liana/permissions_checker.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    end
+
+    def date_difference_in_seconds(date1, date2)
+      ((date1 - date2) * 24 * 60 * 60).to_i
+    end
+
+    def is_permission_expired
+      current_time = DateTime.now
+
+      return true if @@last_retrieve.nil?
```
Comment (arnaudbesnier) [inline app/services/forest_liana/permissions_checker.rb]: why isn't it the first line of the function? What is the point to instantiate `current_time` in this case?

File: app/services/forest_liana/permissions_checker.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      elapsed_seconds = date_difference_in_seconds(current_time, @@last_retrieve)
+      elapsed_seconds >= @expiration_in_seconds
+    end
+
+    def retrieve_permissions_and_check_allowed
+      retrieve_permissions
+      is_allowed
+    end
+
+    def is_authorized
```
Comment (arnaudbesnier) [inline app/services/forest_liana/permissions_checker.rb]: Rails (or Ruby) convention is to put a question mark character `?` at the end of functions that return a boolean.

File: app/controllers/forest_liana/resources_controller.rb [hunk 4]
```diff
@@ -15,6 +15,14 @@ class ResourcesController < ForestLiana::ApplicationController
 
     def index
       begin
+        if request.format == 'csv'
+          return head :forbidden if !PermissionsChecker.new(@resource, 'export').is_authorized
```
Comment (arnaudbesnier) [inline app/controllers/forest_liana/resources_controller.rb]: `if !` => `unless`

File: app/services/forest_liana/permissions_checker.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      if is_permission_expired
+        return retrieve_permissions_and_check_allowed
+      end
+
+      if !is_allowed
+        return retrieve_permissions_and_check_allowed
+      end
```
Comment (arnaudbesnier) [inline app/services/forest_liana/permissions_checker.rb]:
```ruby
if is_permission_expired || !is_allowed
  return retrieve_permissions_and_check_allowed
end
```

File: app/services/forest_liana/permissions_getter.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+          request = Net::HTTP::Get.new(@uri.path)
+          request['Content-Type'] = 'application/json'
+          request['forest-secret-key'] = ForestLiana.env_secret
+          request['forest-token'] = @forest_token if @forest_token
+          response = client.request(request)
+
+          handle_service_response(response)
+        end
+      rescue => exception
+        puts exception
```
Comment (arnaudbesnier) [inline app/services/forest_liana/permissions_getter.rb]: Do we want `puts` in our code?

File: app/services/forest_liana/permissions_checker.rb (hunk 1 above)
Comment (ArnaudValensi) [reply app/services/forest_liana/permissions_checker.rb]: Your example does not convert the env variable to int

File: app/services/forest_liana/users_getter.rb [hunk 7]
```diff
@@ -1,37 +1,19 @@
...
-
-          handle_service_response(response)
-        end
-      rescue => exception
-        puts exception
-        FOREST_LOGGER.error "Cannot retrieve any users for the project " \
-          "you\'re trying to unlock. Forest API seems to be down right now."
-        nil
-      end
+      @preform_request
```
Comment (arnaudbesnier) [inline app/services/forest_liana/users_getter.rb]: Where is this variable defined?

File: app/services/forest_liana/permissions_checker.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    @@permissions = nil
+    @@last_retrieve = nil
+
+    def initialize(resource, permission_name)
+      @collection_name = ForestLiana.name_for(resource)
+      @permission_name = permission_name
+
+      @@expiration_in_seconds = !ENV['FOREST_PERMISSIONS_EXPIRATION_IN_SECONDS'].nil? ?
+        ENV['FOREST_PERMISSIONS_EXPIRATION_IN_SECONDS'].to_i :
+        3600
```
Comment (arnaudbesnier) [inline app/services/forest_liana/permissions_checker.rb]: `@expiration_in_seconds = (ENV['FOREST_PERMISSIONS_EXPIRATION_IN_SECONDS'] || 3600).to_i`

Comment (arnaudbesnier) [reply app/services/forest_liana/permissions_checker.rb]: Why to instantiate this here? You can do it out of the `initialize` method (at line 7 for example)

File: app/services/forest_liana/permissions_checker.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    end
+
+    def date_difference_in_seconds(date1, date2)
+      ((date1 - date2) * 24 * 60 * 60).to_i
+    end
+
+    def is_permission_expired?
+      return true if @@last_retrieve.nil?
+
+      current_time = DateTime.now
```
Comment (arnaudbesnier) [inline app/services/forest_liana/permissions_checker.rb]: You can use `Time.now` without `require 'date'`

File: app/services/forest_liana/permissions_checker.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+      return @@permissions[@collection_name]["collection"][@permission_name];
+    end
+
+    def retrieve_permissions
+      @@permissions = PermissionsGetter.new.perform
+      @@last_retrieve = DateTime.now
+    end
+
+    def date_difference_in_seconds(date1, date2)
+      ((date1 - date2) * 24 * 60 * 60).to_i
```
Comment (arnaudbesnier) [inline app/services/forest_liana/permissions_checker.rb]: using `Time.now` this function can be simplified to `(date1 - date2).to_i`

File: app/services/forest_liana/users_getter.rb (hunk 7 above)
Comment (arnaudbesnier) [reply app/services/forest_liana/users_getter.rb]: This service is broken, this is a major issue.

**Pull request merged**